Overhaul G_EXTRACT_VECTOR_ELT legalization strategy

konstantinschwarz opened the pull request:
We now legalize all `G_AIE_[ZS]EXT_EXTRACT_VECTOR_ELT/G_EXTRACT_VECTOR_ELT` to have 512-bit source vectors.
This allows us to delete a lot of redundant instruction selection code and makes the code much more re-usable across AIE architectures.

Best to review commit by commit, each one should be self-contained.

konstantinschwarz committed:
[AIEX] Pre-commit LLVM IR, legalization tests for extractelement

konstantinschwarz committed:
[AIEX] Overhaul G_EXTRACT_VECTOR_ELT legalization strategy.

We already custom legalized every G_EXTRACT_VECTOR_ELT into G_AIE_SEXT_VECTOR_ELT.
This change ensures we are only producing legal G_AIE_SEXT_VECTOR_ELT by re-using
the previously introduced legalization action.

konstantinschwarz committed:
[AIEX] Introduce and use emitPadVector in AIELegalizerHelper

We start to duplicate similar code across different legalization functions.

konstantinschwarz committed:
[AIEX] Legalize G_AIE_[ZS]EXT_EXTRACT_VECTOR_ELT to legal 512-bit inputs

G_AIE_[ZS]EXT_EXTRACT_VECTOR_ELT are typically introduced in the pre-legalizer combiner.
Since target specific opcodes are considered legal by the Legalizer pass, we need to
pre-legalize these instructions in the combiner.

konstantinschwarz committed:
[AIE] Disable matchExtractVecEltAndExt combine for vector sources <= 64-bit

Keeping these as generic G_EXTRACT_VECTOR_ELTs allows us to use the existing bitcast legalization,
lowering such extracts into bit-arithmetic.

konstantinschwarz committed:
[AIEX] Don't generate G_AIE_[SZ]EXT_EXTRACT_VECTOR_ELT in matchShuffleToExtractBroadcast

This is in preparation of only building legal sized (512-bit) G_AIE_[SZ]EXT_EXTRACT_VECTOR_ELT.

konstantinschwarz committed:
[AIEX] Enforce only legal G_AIE_[SZ]EXT_EXTRACT_VECTOR_ELT

This adds MachineVerifier support for G_AIE_[SZ]EXT_EXTRACT_VECTOR_ELT to enforce only 512-bit source vectors are used after legalization.

konstantinschwarz committed:
[AIEX] Pad inputs to G_AIE_SEXT_EXTRACT_VECTOR_ELT to 512-bit vectors in G_FADD/FSUB legalization

konstantinschwarz committed:
[AIE2P] Add instruction selection support for G_CONCAT_VECTORS of 2 <4 x s64>

martien-de-jong commented on llvm/lib/Target/AIE/aie2p/AIE2PInstructionSelector.cpp:
```diff
@@ -622,6 +622,21 @@ bool AIE2PInstructionSelector::selectG_CONCAT_VECTORS(
     return true;
```
Two returns in a row?

konstantinschwarz replied:
🤔 removed

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -33,6 +33,26 @@ const AIEBaseInstrInfo *AIELegalizerHelper::getInstrInfo() {
+
+  const LLT OrigTy = MRI.getType(SrcReg);
+  const unsigned NumPadElts =
+      WideTy.getSizeInBits() / OrigTy.getSizeInBits() - 1;
```
assert that this divides exactly?

martien-de-jong commented on llvm/test/CodeGen/AIE/GlobalISel/legalize-vsel.mir:
```diff
@@ -21,7 +21,8 @@ body:             |
     ; CHECK-NEXT: [[CONCAT_VECTORS:%[0-9]+]]:_(<64 x s8>) = G_CONCAT_VECTORS [[DEF]](<32 x s8>), [[COPY1]](<32 x s8>)
-    ; CHECK-NEXT: [[CONCAT_VECTORS1:%[0-9]+]]:_(<64 x s8>) = G_CONCAT_VECTORS [[DEF1]](<32 x s8>), [[COPY1]](<32 x s8>)
+    ; CHECK-NEXT: [[COPY2:%[0-9]+]]:_(<32 x s8>) = COPY [[DEF]](<32 x s8>)
+    ; CHECK-NEXT: [[CONCAT_VECTORS1:%[0-9]+]]:_(<64 x s8>) = G_CONCAT_VECTORS [[DEF1]](<32 x s8>), [[COPY2]](<32 x s8>)
```
Perhaps factor out an NFC commit? Or doesn't that work. I can imagine the factored code starts out better.

konstantinschwarz replied:
I considered this commit already a preparatory commit. Not sure what can be factor out in addition here?

martien-de-jong commented on llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp:
Hmm. Sounds like a flaw in the legalizer that we should legalize here. Mais soit.

konstantinschwarz replied:
Target specific instruction do not have associated LegalizerInfo actions, because the generic LegalizerHelper doesn't know anything about them.
An alternative would be to *always* call a target specific custom legalization function for any target specific instruction. I felt the current solution is reasonable

martien-de-jong commented on llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp:
```diff
@@ -284,6 +286,16 @@ bool AIE2PreLegalizerCombinerImpl::tryCombineAll(MachineInstr &MI) const {
+  case AIE2::G_AIE_ZEXT_EXTRACT_VECTOR_ELT:
+  case AIE2::G_AIE_SEXT_EXTRACT_VECTOR_ELT: {
+    const LLT SrcVecTy = MRI.getType(MI.getOperand(1).getReg());
+    if (SrcVecTy.getSizeInBits() != 512) {
```
Would it make sense to have a BasicVectorBitSize in a base class?

gbossu replied:
Or maybe just a static attribute within `AIE2PreLegalizerCombinerImpl`

F-Stuckmann replied:
we have the magic number 512 in multiple files...

konstantinschwarz replied:
We recently introduced `getExtractSubvecNativeSrcSize` in `AIEBaseInstrInfo.h`.
I guess I could rename that function to `getBasicVectorBitSize()` and use it in existing places and the change here.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -33,6 +36,18 @@ const AIEBaseInstrInfo *AIELegalizerHelper::getInstrInfo() {
+  assert(MI.getOpcode() == TargetOpcode::G_UNMERGE_VALUES);
+
+  const int StartIdx = Regs.size();
+  const int NumResults = MI.getNumOperands() - 1;
```
Isn't there an MI.defs() that we could iterate over?

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -33,6 +36,18 @@ const AIEBaseInstrInfo *AIELegalizerHelper::getInstrInfo() {
+  const int StartIdx = Regs.size();
+  const int NumResults = MI.getNumOperands() - 1;
+  Regs.resize(Regs.size() + NumResults);
+  for (int I = 0; I != NumResults; ++I)
```
why not < NumResults?

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1372,4 +1387,105 @@ bool AIELegalizerHelper::legalizeBinOp(LegalizerHelper &Helper,
+static bool fewerElementsAIEVectorExtractElt(MachineRegisterInfo &MRI,
+                                             MachineIRBuilder &MIRBuilder,
+                                             LegalizerHelper &Helper,
+                                             MachineInstr &MI) {
```
Add a brief description? what is MI? fewer than what?

konstantinschwarz replied:
Done. This now also takes `LegalVectorSize` that can be set to a target specific size.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1372,4 +1387,105 @@ bool AIELegalizerHelper::legalizeBinOp(LegalizerHelper &Helper,
+
+  // If the index is a constant, we can really break this down as you would
+  // expect, and index into the target size pieces.
+  int64_t IdxVal;
```
I think IdxVal could be local const initialized in the scope below.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1372,4 +1387,105 @@ bool AIELegalizerHelper::legalizeBinOp(LegalizerHelper &Helper,
+  if (MaybeCst) {
+    IdxVal = MaybeCst->Value.getSExtValue();
+    // Avoid out of bounds indexing the pieces.
+    if (IdxVal >= VecTy.getNumElements()) {
```
or < 0?

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1372,4 +1387,105 @@ bool AIELegalizerHelper::legalizeBinOp(LegalizerHelper &Helper,
+    // Avoid out of bounds indexing the pieces.
+    if (IdxVal >= VecTy.getNumElements()) {
+      MIRBuilder.buildUndef(DstReg);
+      return true;
```
I guess we could move this check up. It's not related to the narrow vector type at all.

konstantinschwarz replied:
Right, I moved the code breaking the vector into sub-vectors after the check for out-of-bounds.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1372,4 +1387,105 @@ bool AIELegalizerHelper::legalizeBinOp(LegalizerHelper &Helper,
+    LLT IdxTy = MRI.getType(Idx);
+    int64_t PartIdx = IdxVal / NewNumElts;
+    auto NewIdx =
+        MIRBuilder.buildConstant(IdxTy, IdxVal - NewNumElts * PartIdx);
```
nit: I think it's natural to write this as `IdxVal % NewNumElts`

konstantinschwarz replied:
I "borrowed" the constant idx handling code from the generic LegalizerHelper, which also handles odd types.
Here we indeed only expect to see types of exact multiples.
I'll simplify accordingly.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1372,4 +1387,105 @@ bool AIELegalizerHelper::legalizeBinOp(LegalizerHelper &Helper,
+  unsigned NewNumElts = NarrowVecTy.getNumElements();
+
+  LLT IdxTy = MRI.getType(Idx);
+  auto Cst =
```
More descriptive names for Cst, Cst2?
Most of these decls could be const.

konstantinschwarz replied:
Done

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1372,4 +1387,105 @@ bool AIELegalizerHelper::legalizeBinOp(LegalizerHelper &Helper,
+      MIRBuilder.buildConstant(IdxTy, (VecParts.size() - 2) * NewNumElts);
+  Register AdjImm = MIRBuilder.buildSelect(IdxTy, Cmp, Cst2, Cst).getReg(0);
+
+  for (size_t I = VecParts.size() - 2; I > 0; I--) {
```
Can't we roll the code above into this loop? It looks very similar.

konstantinschwarz replied:
hm indeed, I thought it would complicate the loop more, but this looks better!

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1372,4 +1387,105 @@ bool AIELegalizerHelper::legalizeBinOp(LegalizerHelper &Helper,
+  const LLT SrcVecTy = MRI.getType(SrcVecReg);
+  const Register IdxReg = MI.getOperand(2).getReg();
+
+  assert(SrcVecTy.getSizeInBits() != 512 && "Instruction already legal!");
```
Why bother? just return true?

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1372,4 +1387,105 @@ bool AIELegalizerHelper::legalizeBinOp(LegalizerHelper &Helper,
+      return false;
+  } else {
+    llvm_unreachable(
+        "Illegal vector size for G_AIE_[ZS]EXT_EXTRACT_VECTOR_ELT");
```
This catches 544, but not 1056

konstantinschwarz replied:
Added a check for power-of-2

martien-de-jong commented on llvm/lib/Target/AIE/AIE2LegalizerInfo.cpp:
```diff
@@ -405,21 +405,15 @@ AIE2LegalizerInfo::AIE2LegalizerInfo(const AIE2Subtarget &ST) : AIEHelper(ST) {
       })
-      // If it is 32-bit, the LLVM can perform some bitshifts to legalize it
+      // Bitcast 32-bit vectors to 1 x s32, which will be expanded to bit
+      // arithmetic
```
Would this also work for 64 bit?

konstantinschwarz replied:
I think so, but I'd like to avoid packing even more changes into this PR

martien-de-jong commented on llvm/lib/Target/AIE/AIE2LegalizerInfo.cpp:
```diff
@@ -428,11 +422,15 @@ AIE2LegalizerInfo::AIE2LegalizerInfo(const AIE2Subtarget &ST) : AIEHelper(ST) {
+      .clampMinNumElements(1, S32, 8)
+      // Custom legalize resulting vector types >= 256-bit
+      .customIf(typeInSet(1, {V8S32, V16S32, V32S32, V16S16, V32S8, V32S16,
+                              V64S8, V64S16, V128S8}));
```
nit: strange grouping of this set. Comment would suggest grouping by size.

martien-de-jong commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -664,26 +664,37 @@ bool AIELegalizerHelper::legalizeG_EXTRACT_VECTOR_ELT(LegalizerHelper &Helper,
             SrcVecEltTy == S32) &&
            "Unexpected vector element type for extract vector elt!");
+
+    MachineInstr *ExtractInstr;
```
Please nullptr initialize

martien-de-jong commented on llvm/lib/Target/AIE/aie2p/AIE2PLegalizerInfo.cpp:
```diff
@@ -483,11 +471,19 @@ AIE2PLegalizerInfo::AIE2PLegalizerInfo(const AIE2PSubtarget &ST)
+      .customIf(typeInSet(1, AIE2PAccumulatorTypes))
+      .customIf([=](const LegalityQuery &Query) {
+        const LLT &VectTy = Query.Types[1];
+        return VectTy.getElementType().getSizeInBits() >= 256;
```
We can't do this for AIE2?

konstantinschwarz replied:
We can, but wasn't needed before. Additional changes should come in a separate PR

gbossu commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/inst-select-concat-vectors.mir:
```diff
@@ -8,6 +8,27 @@
+    %1:vregbank(<4 x s64>) = G_IMPLICIT_DEF
+    %2:vregbank(<8 x s64>) = G_CONCAT_VECTORS %0(<4 x s64>), %1(<4 x s64>)
+    PseudoRET implicit $lr, implicit %2
+...
```
Is there a corresponding legalizer test?

konstantinschwarz replied:
No, we have very little G_CONCAT_VECTOR legalization test coverage. There's also a nice opportunity to move instruction selection code to the legalizer and make it common across architectures.
I'll add the missing legalizer test to this commit, but the re-work of G_CONCAT_VECTOR legalization strategy should come in a separate PR

gbossu commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -1877,13 +1877,12 @@ bool llvm::matchShuffleToExtractBroadcast(MachineInstr &MI,
-    auto Extr = B.buildInstr(ExtractOpc, {LLT::scalar(32)}, {SrcVecReg, Cst});
+    const LLT DstElemTy = MRI.getType(SrcVecReg).getElementType();
+    auto Extr =
+        B.buildExtractVectorElementConstant(DstElemTy, SrcVecReg, *UniqOpIdx);
```
Nit: Update the docstring for this function, it does not mention `G_EXTRACT_VECTOR_ELT`

konstantinschwarz replied:
done

martien-de-jong commented on llvm/lib/Target/AIE/AIE2InstrInfo.cpp:
```diff
@@ -153,6 +153,15 @@ bool AIE2InstrInfo::verifyGenericInstruction(const MachineInstr &MI,
+    if (MRI.getType(MI.getOperand(1).getReg()).getSizeInBits() != 512) {
+      bool IsLegalized =
+          MI.getParent()->getParent()->getProperties().hasProperty(
+              MachineFunctionProperties::Property::Legalized);
```
How about a static bool AIEBaseInstrInfo::isLegalized(MI) ?

konstantinschwarz replied:
Good idea, added

martien-de-jong commented on llvm/test/CodeGen/AIE/verifier/verify-szext-extract-vec-elt.mir:
```diff
@@ -16,26 +16,32 @@ body:             |
-    %2:_(s32) = G_AIE_SEXT_EXTRACT_VECTOR_ELT %0(<16 x s16>), %1(s32)
-    %3:_(s32) = G_AIE_ZEXT_EXTRACT_VECTOR_ELT %0(<16 x s16>), %1(s32)
+    %2:_(s32) = G_AIE_SEXT_EXTRACT_VECTOR_ELT %0(<32 x s16>), %1(s32)
+    %3:_(s32) = G_AIE_ZEXT_EXTRACT_VECTOR_ELT %0(<32 x s16>), %1(s32)
```
CHECK: these tests had illegal input

konstantinschwarz replied:
After the MachineVerifier change, we consider these illegal. I added these tests (< 512-bit vectors) to the `nok` test

gbossu commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -33,6 +33,26 @@ const AIEBaseInstrInfo *AIELegalizerHelper::getInstrInfo() {
+  for (unsigned I = 0; I < NumPadElts; I++)
+    Regs.push_back(UndefReg);
+  const Register NewSrcReg = MRI.createGenericVirtualRegister(WideTy);
+  MIRBuilder.buildMergeLikeInstr(NewSrcReg, Regs);
```
Oh interesting! I guess this does either a `G_CONCAT_VECTORS` or `G_MERGE_VALUES`?

konstantinschwarz replied:
Yes, depending on whether we supply scalars (G_MERGE_VALUES) or vectors (G_CONCAT_VECTORS).

gbossu commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/legalize-extract-vector-elt.mir:
```diff
@@ -301,7 +301,8 @@ body:             |
     ; CHECK-NEXT: [[COPY2:%[0-9]+]]:_(<8 x s32>) = COPY [[DEF1]](<8 x s32>)
     ; CHECK-NEXT: [[CONCAT_VECTORS:%[0-9]+]]:_(<16 x s32>) = G_CONCAT_VECTORS [[BITCAST]](<8 x s32>), [[COPY2]](<8 x s32>)
-    ; CHECK-NEXT: [[CONCAT_VECTORS1:%[0-9]+]]:_(<16 x s32>) = G_CONCAT_VECTORS [[BITCAST1]](<8 x s32>), [[COPY2]](<8 x s32>)
+    ; CHECK-NEXT: [[COPY3:%[0-9]+]]:_(<8 x s32>) = COPY [[DEF1]](<8 x s32>)
```
Why do we now need an extra copy?

konstantinschwarz replied:
Previously we built `MIRBuilder.buildUndef(UndefReg);` in `legalizeG_SELECT` and re-used it for padding both src1 and src2.
Now we build the Undef within `emitPadUndefVector`, which we call twice. CSEMIRBuilder still folds the build to a COPY of the already existing Undef.

martien-de-jong commented on llvm/lib/Target/AIE/aie2p/AIE2PInstrPatterns.td:
```diff
@@ -1096,3 +1096,14 @@ foreach Ty = [v64i8, v32i16, v16i32, v8i64] in {
+defm : Extract_512<i32, v64i8, (i32 c6u:$idx), VEXTRACT_8_vec_extract_imm_vaddSign0, VEXTRACT_8_vec_extract_imm_vaddSign1>;
+defm : Extract_512<i32, v32i16, (i32 c6u:$idx), VEXTRACT_16_vec_extract_imm_vaddSign0, VEXTRACT_16_vec_extract_imm_vaddSign1>;
+defm : Extract_512<i32, v16i32, (i32 c6u:$idx), VEXTRACT_32_vec_extract_imm_vaddSign0, VEXTRACT_32_vec_extract_imm_vaddSign1>;
+defm : Extract_512<i64, v8i64, (i32 c6u:$idx), VEXTRACT_64_vec_extract_imm_vaddSign0, VEXTRACT_64_vec_extract_imm_vaddSign1>;
```
CHECK: number of index bits is independent of element size

martien-de-jong commented on llvm/test/CodeGen/AIE/GlobalISel/inst-select-extract-vector-elem.mir:
```diff
@@ -30,9 +30,8 @@ body:             |
-    ; AIE2P-NEXT: [[VEXTRACT_16_vec_extract_r_vaddSign0_:%[0-9]+]]:er = VEXTRACT_16_vec_extract_r_vaddSign0 [[COPY]], [[MOV_RLC_imm11_pseudo]], implicit $vaddsign0
-    ; AIE2P-NEXT: $r0 = COPY [[VEXTRACT_16_vec_extract_r_vaddSign0_]]
+    ; AIE2P-NEXT: [[VEXTRACT_16_vec_extract_imm_vaddSign0_:%[0-9]+]]:er = VEXTRACT_16_vec_extract_imm_vaddSign0 [[COPY]], 1, implicit $vaddsign0
+    ; AIE2P-NEXT: $r0 = COPY [[VEXTRACT_16_vec_extract_imm_vaddSign0_]]
```
Check: we gained  an immediate mode on AIE2P

konstantinschwarz replied:
Yes, those are new on AIE2P, and significantly reduce the register pressure on GPRs for constant idx extracts

martien-de-jong reviewed (commented):
Oh, I have far too many comments for such an heroic effort. In general it looks good to me. I would mainly like to avoid uninitialized locals, and have const on declarations that have longer scope than just a few lines.

andcarminati commented on llvm/test/CodeGen/AIE/GlobalISel/legalize-extract-vector-elt.mir:
```diff
@@ -8,6 +8,693 @@
 # RUN: llc -mtriple aie2 -run-pass=legalizer %s -verify-machineinstrs -o - | FileCheck %s
 # RUN: llc -mtriple aie2p -run-pass=legalizer %s -verify-machineinstrs -o - | FileCheck %s
 
+---
```
nit: 2023-2025

andcarminati commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/prelegalizercombiner-extract-vector-elt.mir:
```diff
@@ -0,0 +1,377 @@
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
```
nit: 2025.

andcarminati commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/inst-select-concat-vectors.mir:
```diff
@@ -8,6 +8,27 @@
 # RUN: llc -mtriple aie2p -run-pass=instruction-select %s -verify-machineinstrs -o - | FileCheck %s
 
+---
+name:            vconcat_512_acc64
```
I think this name can cause some confusion because of acc. Maybe just  `vconcat_512`

konstantinschwarz replied:
Changed it to vconcat_v8s64_v4s64

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -995,6 +995,10 @@ bool llvm::matchExtractVecEltAndExt(
   LLT SrcVecTy = MRI.getType(MI.getOperand(1).getReg());
+  // Extracts from vectors <= 64-bits are lowered to bit-arithmetic in
+  // legalization
+  if (SrcVecTy.getSizeInBits() <= 64)
```
Can we extend the test to cover size=64?

konstantinschwarz replied:
I added the tests for 64-bit vectors to `llvm/test/CodeGen/AIE/GlobalISel/prelegalizercombiner-extract-vector-elt.mir` in the first commit, they now show the new behavior in this commit.

Note however that we do not support 64-bit src operands in the legalizer yet. I don't want to add even more changes to this PR, handling all missing vector sizes should come in a separate PR.

F-Stuckmann commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1211,27 +1260,17 @@ bool AIELegalizerHelper::legalizeG_SELECT(LegalizerHelper &Helper,
   MIRBuilder.buildInstr(MI.getOpcode(), {NewDstReg},
                         {SrcReg0, NewSrcReg1, NewSrcReg2}, MI.getFlags());
 
+  const unsigned NumPadElts = (512 / DstVecSize) - 1;
```
Maybe add a comment describing the magic number 512

martien-de-jong replied:
Nah, give it a sensible name and get it from TII.

konstantinschwarz replied:
I'll just use the already existing `MaxBitSize`, missed it before

F-Stuckmann commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1376,4 +1401,105 @@ bool AIELegalizerHelper::legalizeBinOp(LegalizerHelper &Helper,
+    unsigned NewNumElts = NarrowVecTy.getNumElements();
+
+    LLT IdxTy = MRI.getType(Idx);
+    int64_t PartIdx = IdxVal / NewNumElts;
```
If we don't enter `if (MaybeCst) ` IdxVal won't be properly set

F-Stuckmann commented on llvm/lib/Target/AIE/AIELegalizerHelper.cpp:
```diff
@@ -1376,4 +1401,105 @@ bool AIELegalizerHelper::legalizeBinOp(LegalizerHelper &Helper,
+  int64_t IdxVal;
+  auto MaybeCst = getIConstantVRegValWithLookThrough(Idx, MRI);
+  if (MaybeCst) {
+    IdxVal = MaybeCst->Value.getSExtValue();
```
How do we handle ZEXT?

konstantinschwarz replied:
This is getting the immediate from a G_CONSTANT operand, which we want to sign-extend. There is no ZEXT case to handle here

konstantinschwarz committed:
[NFC][AIEX] Rename getExtractSubvecNativeSrcSize to getBasicVectorBitSize

konstantinschwarz reviewed (commented):
Thanks a lot for the reviews!
I think I addressed most/all of the comments. In particular, got rid of the hard-coded legal vector size

martien-de-jong commented on llvm/lib/Target/AIE/AIEBaseInstrInfo.h:
```diff
@@ -482,6 +482,13 @@ struct AIEBaseInstrInfo : public TargetInstrInfo {
 
+  /// Returns whether the function that contains \p MI has been
+  /// legalized
+  bool isLegalized(const MachineInstr &MI) const {
```
nit: could be static here

konstantinschwarz replied:
done

martien-de-jong reviewed (dismissed):
Nothing blocking

konstantinschwarz committed:
[AIE2][AIE2P] Move selection of G_AIE_[ZS]EXT_EXTRACT_VECTOR_ELT to TableGen patterns

For AIE2, this is NFC. For AIE2P, we now benefit from patterns supporting constant indices.

niwinanto commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -1863,7 +1863,8 @@ static std::optional<int> getUniqueIndex(ArrayRef<int> Mask) {
 ///         %2:_(<8 x s64>) = G_SHUFFLE_VECTOR %X(<4 x s64>), %1(<4 x s64>),
 ///         shufflemask(3, 3, 3, 3, 3, 3, 3, 3)
-/// To :    %2:_(<8 x s64>) = G_AIE_BROADCAST_VECTOR %X(<4 x s64>)
+/// To :    %3:_(s64) = G_EXTRACT_VECTOR_ELT %X, 3
```
@konstantinschwarz It seems we are not checking the extract element belongs to first source element. Canonicalization combine from the upstream might solve this, but till then this combine produces wrong results. For example, if shuffle mask in the doc string is (4, 4, ...)

niwinanto reviewed (commented):
Really nice work. Just one comment, not really part of this, but found while reviewing this.
